### PR TITLE
pruntime: Unblock concurrent requests

### LINKED ISSUE #1197: Concurrent query issues in pRuntime
During my recent testing, I found two concurrency issues with pRuntime:

- [ ]  When a query is blocked inside the contract, the next query will get stuck (unable to enter the rocket).
    
    We didn't anticipate this problem before. Even though we have 8 rocket threads, only two concurrent requests can cause the blockage. We thought that Tokio's work-stealing mechanism would distribute the waiting tasks in the blocked thread to other threads, but in fact, it doesn't.
    
    This problem can be solved by using tokio::spawn_blocking().
    
- [ ]  When pRuntime is synchronizing blocks and executing transactions, queries cannot be processed because Phactory is locked.
    This is a known issue. When executing transactions, the global Phactory instance is locked. Almost all of the pRPC requests lock Phactory to retrieve the state, including queries. Most RPC release Phacktory quickly after retrieving the required state because it's read-only. However, executing transactions requires modifying the state, which locks it for a long duration, making it impossible to process queries(and any other requests) at the same time.
    To solve this problem, we may need to organize and isolate the states in Phactory, use some kind of RCU lock mechanism, or change to a fine-grained lock to minimize the locking time of transactions.

<details>
  <summary>Chinese</summary>
  <p>
我这两天测试发现pRuntime存在的两个并发问题：

- 当一个query阻塞在contract里面后，下一个query会卡住（无法进入rocket）
    
    这个问题之前没有预料到，即使我们开8个rocket线程，仅仅两个并发请求就会卡住第二个。一直认为tokio的work stealing机制会将阻塞线程中的堆积任务分派到其它线程去，事实上并不能。
    
    此问题可以使用tokio::spawn_blocking()解决。
    
- 当pruntime正在同步区块执行交易时，因为Phactory被加锁而无法处理query
    
    这个是已知问题。当执行交易时，会将全局Phactory实例锁住，会更新里面较多的状态。我们几乎所有请求都会锁Phactory取状态，query也会锁，只是取出需要的状态后很快释放，因为它是只读的。而执行交易需要修改状态，会加锁较长时间，导致无法同时处理query。
    
    要解决这个问题可能需要对Phactory中的状态进行梳理隔离，通过某种RCU锁机制，或改成细粒度锁，尽量缩短tx的加锁时间
  </p>
</details>

### PR DESCRIPTION
**Pull request by kvinwang**
This PR fixes #1197 

# How it works
This is achieved by using an RCU-like mechanism. When a chain block is dispatched to a worker, pruntime copies the entire Phactory state, and processes the block with the copied Phactory. When the block processing is finished, The old Phactory would be replaced with the copied new one. This enables all read on the old Phactory while processing blocks.

To make the heavy copying as cheap as possible, we replaced most of the container types such as `BTreeMap` with the equivalent ones from `im::*`. And we DON'T do the RCU while syncing history blocks.

### CODE REVIEW

File: crates/phactory/src/prpc_service.rs [hunk 1]
```diff
@@ -1139,12 +1167,12 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> PhactoryApi for Rpc
...
         req: pb::GetRuntimeInfoRequest,
     ) -> RpcResult<pb::InitRuntimeResponse> {
-        self.lock_phactory()
+        self.lock_phactory(true)?
             .get_runtime_info(req.force_refresh_ra, req.decode_operator()?)
     }
 
     async fn get_egress_messages(&mut self, _: ()) -> RpcResult<pb::GetEgressMessagesResponse> {
-        self.lock_phactory()
+        self.lock_phactory(false)?
```
Comment (h4x3rotab) [inline crates/phactory/src/prpc_service.rs:1175]: Why this one is `false`?

**Review by h4x3rotab — COMMENTED** [review]
Does the massive `im::OrdMap` to `BTreeMap` replacement introduce overhead?

File: crates/phactory/src/prpc_service.rs [hunk 2]
```diff
@@ -1118,14 +1116,44 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> PhactoryApi for Rpc
     /// Dispatch blocks (Sync storage changes)"
     async fn dispatch_blocks(&mut self, request: pb::Blocks) -> RpcResult<pb::SyncedTo> {
         let blocks = request.decode_blocks()?;
-        self.lock_phactory().dispatch_block(self.req_id, blocks)
+        let mut phactory = {
```
[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/phactory/src/prpc_service.rs:1119]: Added some comments there.

**Comment by kvinwang** [comment]
> Does the massive im::OrdMap to BTreeMap replacement introduce overhead?

Saying syncing speed, using OrdMap is as fast as using BTreeMap and almost the same memory cost.

**Pull request merged**